Add dependencycheck tool to address long running no-vendor-cycles test

hasheddan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup
> /kind flake

**What this PR does / why we need it**:

Adds a new tool, `dependencycheck`, to detect cyclical dependencies on main and staging repos in vendored packages. This is currently being handled by `hack/verify-no-vendor-cycles.sh`, which takes exceedingly long to run (3-4 minutes) for the task it is accomplishing. `dependencycheck` typically runs in a few seconds. In addition, when taking a look at the logs for `verify-master` and trying it out locally, I am not certain that this script is actually accomplishing its goal. For instance, if a package depended on `k8s.io/apimachinery/pkg/runtime`, it appears it is being printed as such, yet the script is checking for a match of `k8s.io/kubernetes/vendor/k8s.io` then `k8s.io/apimachinery`. I am guessing this may have to do with dependencies formerly being printed by their full path, but I am not sure. See more here: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-verify-master/1287005337945116672/build-log.txt

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref: https://github.com/kubernetes/kubernetes/issues/87807 https://github.com/kubernetes/kubernetes/issues/92937

**Special notes for your reviewer**:

I wasn't sure who to add to the `OWNERS` file here so I just picked some of the usual suspects.

/cc @BenTheElder @dims @liggitt 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hasheddan commented:
/sig testing
/priority important-soon

hasheddan commented:
/test pull-kubernetes-files-remake

hasheddan commented:
/test pull-kubernetes-e2e-kind

hasheddan commented:
/retest
`verify.no-vendor-cycles` did not get run that time because of the spike in other tests

BenTheElder commented on cmd/vendorcycle/vendorcycle.go:
```diff
@@ -0,0 +1,95 @@
+			return err
+		}
+		if info.IsDir() {
+			p, err := build.Default.Import(fmt.Sprintf("./%s", path), ".", 0)
```
fun thought: build tags?
I _think_ we're fine, the default tags should get setup in the script below, and non-default build tags shouldn't be a problem, probably?

hasheddan replied:
That is a valid point. I'm not sure how often we will hit it in practice, but I think regardless it might be better to refactor this to use https://godoc.org/golang.org/x/tools/go/packages

hasheddan replied:
Interesting observation, using `packages`, even with only `NeedImports` mode set is _significantly_ slower, which makes me think that it is actually pulling the modules rather than inspecting the source (this may have already been known by others :sweat_smile:). If we can afford to only inspect the source rather than loading the packages then that would certainly be advantageous as demonstrated by the 2 second `verify.no-vendor-cycles` run on the successful `verify-master` here.

liggitt commented:
Thanks for jumping on this, much appreciated. I need to double check whether the current impl detects module level transitive dependencies on k8s.io/kubernetes or staging repos even if the particular packages we use don't depend back on these repos. I think the source scanning technique would miss those. If we currently catch/prevent those, we shouldn't regress that. I'll queue this up to look at early next week.

hasheddan commented:
@liggitt thanks! I _think_ that `go list` will only check package level transitive dependencies, but I'll test and make sure. Regardless of whether it does, I think an adjustment is needed here to make this check "correct" even if we are unable to improve its latency for the time-being.

dims commented:
/uncc

glad to see @liggitt on this!

liggitt commented:
> I need to double check whether the current impl detects module level transitive dependencies on k8s.io/kubernetes or staging repos even if the particular packages we use don't depend back on these repos. I think the source scanning technique would miss those. If we currently catch/prevent those, we shouldn't regress that. I'll queue this up to look at early next week.

looks like we only checked packages found in vendor previously

if possible, I'd like to lean on `go list` to do the dependency resolution, with something like this:
```
go list -mod=vendor -test -deps -json ./vendor/... > deps.json
```

A go program to process that deps.json file seems better than bash

I would envision something like:
1. Determine the list of staging repos
2. Run `go list` to capture vendor dependencies
3. Run `dependencycheck deps.json k8s.io/kubernetes` to fail if any package in deps.json depends on k8s.io/kubernetes/...
4. Run `dependencycheck deps.json k8s.io/client-go k8s.io/api k8s.io/apimachinery ...` to fail if any package in deps.json depends on staging modules

As a follow-up, for the module-level cycle checks, we can separately use `go mod graph` to detect/protect against this, forbidding things that match `go mod graph | grep " k8s.io/kubernetes"`, etc

hasheddan commented:
@liggitt do we need to explicitly run 3 & 4 separately? This check will execute quickly, so it is not a huge deal, but we are iterating twice through when it seems like one iteration checking for both would be faster.

liggitt commented:
it's a little easier to treat staging repos and k8s.io/kubernetes separately, but there's not a strong reason to. I expect the tool will take less than a second to run so I doubt calling it twice would matter

RobertKielty commented:
@hasheddan @liggitt  What's the next step here? 

I'm looking at verify-master failures and I just want to understand if there more changes needed on this PR?

hasheddan commented:
@RobertKielty I need to refactor as has been suggested. I should have this up today. However, this is not currently causing the `verify-master` failures. That is due to pinning resource requests and limits, then typecheck exceeding the memory limit. Opening an issue now with artifacts.

RobertKielty commented:
Thanks @hasheddan

hasheddan commented:
@liggitt I have updated and this is ready for another review :+1: I had one question about the module-level dependency checks. If we are checking at the module level (which I have not added yet), do we also need to be checking at the package level (i.e. aren't the module-level dependencies a superset of the packages)?

liggitt commented:
> If we are checking at the module level (which I have not added yet), do we also need to be checking at the package level (i.e. aren't the module-level dependencies a superset of the packages)?

No. Not all our dependencies declare go.mod files, which means their transitive dependencies are not listed in `go mod graph` and we have to drop down to package level traversals.

hasheddan commented:
@liggitt understood! Thanks for the clarification 👍

liggitt commented on hack/verify-no-vendor-cycles.sh:
```diff
@@ -34,21 +34,8 @@ staging_repos_pattern=$(IFS="|"; echo "${staging_repos[*]}")
 
 cd "${KUBE_ROOT}"
 
```
this will create and clean up a temp dir for you:
```
kube::util::ensure-temp-dir()
go list -mod=vendor -test -deps -json ./vendor/... > "${KUBE_TEMP}/deps.json"
```

hasheddan commented:
@liggitt updated with module checks

BenTheElder commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+var (
+	exclude  = flag.String("exclude", `a^`, "skip packages pattern")
```
we should probably specify that this is a regex. does this default make sense?

hasheddan replied:
Agreed that we should specify. The default here should never match.

hasheddan replied:
However, we could just set it to empty string and if exclude is empty string we can return immediately (no reason to check if we don't care what is matched)

liggitt replied:
yeah, leave empty and ignore if unset

BenTheElder commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,96 @@
+	Deps       []string `json:",omitempty"`
+}
+
+func jsonObjsToArray(raw []byte) []byte {
```
might deserve a comment

liggitt commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,96 @@
+
+var (
+	exclude  = flag.String("exclude", `a^`, "skip packages pattern")
+	restrict = flag.String("restrict", `a^`, "restricted dependencies pattern")
```
leave empty and require to be set

liggitt commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,96 @@
+
+	args := flag.Args()
+
+	if len(args) == 0 {
```
```suggestion
	if len(args) != 1 {
```

liggitt commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,96 @@
+
+	violations := map[string][]string{}
+	for _, p := range packages {
+		if !excludePattern.MatchString(p.ImportPath) {
```
nit: continue early and unnest 

```go
if excludePattern != nil && excludePattern.MatchString(p.ImportPath) {
  continue
}
importViolations := []string{}
for _, d := range p.Deps {
  ...
```

hasheddan commented:
@liggitt I believe all comments have been addressed and this is ready for another pass :+1:

hasheddan commented:
/retest

hasheddan commented:
/retest

liggitt commented:
> no-vendor-cycles  6s

down from 5m18s in a recent run 🎉

liggitt commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,108 @@
+	args := flag.Args()
+
+	if len(args) != 1 {
+		log.Fatalf("usage: dependencycheck <json-dep-file>")
```
nit, give an example of how to generate this file (e.g. `go list -mod=vendor -test -deps -json ./vendor/...`)

liggitt commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,108 @@
+			log.Fatalf("Error compiling excluded package regex: %v", err)
+		}
+	}
+	fmt.Println(excludePattern)
```
drop debug logging?

liggitt commented on hack/verify-no-vendor-cycles.sh:
```diff
@@ -34,21 +34,16 @@ staging_repos_pattern=$(IFS="|"; echo "${staging_repos[*]}")
-  exit 1
-fi
+# Check for any module that is not main or staging and depends on main or staging
+go mod graph | grep -vE "^k8s.io\/(kubernetes|${staging_repos_pattern})" | grep -E "\sk8s.io\/(kubernetes|${staging_repos_pattern})" && exit 1
```
this will not output a helpful error if it fails

suggest something like 

```sh
bad_deps=$(go mod graph | grep -vE "^k8s.io\/(kubernetes|${staging_repos_pattern})" | grep -E "\sk8s.io\/(kubernetes|${staging_repos_pattern})" || true)
if [[ -n "${bad_deps}" ]]; then
  echo "Found disallowed dependencies that transitively depend on k8s.io/kubernetes or staging modules:"
  echo "${bad_deps}"
  exit 1
fi
```

liggitt commented on hack/verify-no-vendor-cycles.sh:
```diff
@@ -34,21 +34,16 @@ staging_repos_pattern=$(IFS="|"; echo "${staging_repos[*]}")
 
```
also suggest dropping `set -x;` it's almost completely noise

liggitt commented on hack/verify-no-vendor-cycles.sh:
```diff
@@ -34,21 +34,16 @@ staging_repos_pattern=$(IFS="|"; echo "${staging_repos[*]}")
+go run cmd/dependencycheck/dependencycheck.go -restrict "^k8s\.io/kubernetes/" "${KUBE_TEMP}/deps.json"
+
+# Check for any vendored package that depends on staging
+go run cmd/dependencycheck/dependencycheck.go -restrict "^k8s\.io/(${staging_repos_pattern})(/|$)" "${KUBE_TEMP}/deps.json"
```
I think both of these need to exclude staging repos from the checks:

`-exclude "^k8s\.io/(${staging_repos_pattern})(/|$)"`

liggitt commented on cmd/dependencycheck/dependencycheck.go:
```diff
@@ -0,0 +1,108 @@
+type goPackage struct {
+	Name       string
+	ImportPath string
+	Deps       []string
```
Deps includes all transitive dependencies of this package, which is likely to be enormous

Imports + TestImports + XTestImports is likely to be more useful in identifying the actual problematic import that needs to be dealt with

suggest:
```diff
--- a/cmd/dependencycheck/dependencycheck.go
+++ b/cmd/dependencycheck/dependencycheck.go
@@ -34,9 +34,11 @@ var (
 )
 
 type goPackage struct {
-       Name       string
-       ImportPath string
-       Deps       []string
+       Name         string
+       ImportPath   string
+       Imports      []string
+       TestImports  []string
+       XTestImports []string
 }
 
 func main() {
@@ -86,7 +88,12 @@ func main() {
                        continue
                }
                importViolations := []string{}
-               for _, d := range p.Deps {
+
+               allImports := []string{}
+               allImports = append(allImports, p.Imports...)
+               allImports = append(allImports, p.TestImports...)
+               allImports = append(allImports, p.XTestImports...)
+               for _, d := range allImports {
                        if depsPattern.MatchString(d) {
                                importViolations = append(importViolations, d)
                        }
```

justaugustus commented:
Very cool, Dan!

hasheddan committed:
Add dependencycheck tool to check for dependency cyles in vendored pkgs

dependencycheck verifies that no violating depdendencies exist in pkgs
passed via a JSON file generated from go list.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Update verify-no-vendor-cycles to use dependencycheck

verify-no-vendor-cycles currently routinely takes 3-4 minutes to run due
to the way that it checks for cyclical imports in vendored packages.
Moving to using the new dependencycheck tool makes this test much faster.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan commented:
@liggitt updates complete :+1: However, this is going to fail on `pull-kubernetes-verify` due to https://github.com/kubernetes/kubernetes/issues/93623

liggitt commented:
/lgtm
/approve
/hold for publishing bot fix

liggitt commented:
/test pull-kubernetes-e2e-kind

liggitt commented:
/retest
/hold cancel

BenTheElder commented:
thanks @hasheddan @liggitt
